lib/IOExt/IOExt.cpp

```diff
@@ -3,253 +3,219 @@
+  for 5V bus
+      * Connect a 4.7k resistor between SDA and Vcc
+      * Connect a 4.7k resistor between SCL and Vcc
+  for 3.3V bus
+      * Connect a 2.4k resistor between SDA and Vcc
+      * Connect a 2.4k resistor between SCL and Vcc
+  */
+  printf("Scanning I2C Addresses:\n");
+  uint8_t cnt = 0;
+  for (uint8_t i = 0; i < 0x80; i++) {
+    Wire.beginTransmission(i);
+    uint8_t ec = Wire.endTransmission(true);
+    if (ec == 0) {
+      printf("%02x ", i);
+      cnt++;
+    } else {
+      printf("-- ");
+    }
+    if ((i & 0x0f) == 0x0f) {
+      printf("\n");
+    }
+  }
+  printf("Scan Completed, %d I2C Devices found.\n", cnt);
 }
 
-bool i2cReady(uint8_t adr)
-{
-    uint32_t timeout = millis();
-    bool ready = false;
-    while ((millis() - timeout < 100) && (!ready))
-    {
-        Wire.beginTransmission(adr);
-        ready = (Wire.endTransmission() == 0);
-    }
-    return ready;
+bool i2cReady(uint8_t adr) {
+  uint32_t timeout = millis();
+  bool ready = false;
+  while ((millis() - timeout < 100) && (!ready)) {
+    Wire.beginTransmission(adr);
+    ready = (Wire.endTransmission() == 0);
+  }
+  return ready;
 }
 
-void init_ioext()
-{
-
-    // CRITICAL SECTION I2C: start
-    xSemaphoreTake(i2c_mutex, portMAX_DELAY);
-
-    // setup pins
-    pcf8574.pinMode(P0, OUTPUT, LOW);
-    pcf8574.pinMode(P1, INPUT);
-    pcf8574.pinMode(P2, INPUT);
-    pcf8574.pinMode(P3, INPUT);
-    pcf8574.pinMode(P4, INPUT);
-    pcf8574.pinMode(P5, INPUT);
-    pcf8574.pinMode(P6, INPUT);
-    pcf8574.pinMode(P7, INPUT);
-
-    Wire.begin();
-    scan();
-
-    // start
-    if (pcf8574.begin())
-    {
-        printf("[PCF8574] Init successful.\n");
-    }
-    else
-    {
-        printf("[PCF8574] Init failed.\n");
-    }
-
-    xSemaphoreGive(i2c_mutex);
-    // CRITICAL SECTION I2C: end
-}
-
-//volatile PCF8574::DigitalInput digitalIn;
-volatile bool ioInterruptRequest = false;
-void key_pressed_interrupt_handler()
-{
-    ioInterruptRequest = true;
-    //digitalIn = pcf8574.digitalReadAll();
-    //buf = readBuffer();
-    // Wire.requestFrom(0x20,(uint8_t)1);
-    // ioBuf = Wire.read();
-}
-
-void _speedCheck(int speed)
-{
-    if (speed < 50)
-    {
-        arrow_increase(true);
-    }
-    else
-    {
-        arrow_increase(false);
-    }
-    if (speed > 80)
-    {
-        arrow_decrease(true);
-    }
-    else
-    {
-        arrow_decrease(false);
-    }
-}
+void init_ioext() {
 
-void _handleIoInterrupt()
-{
-    // CRITICAL SECTION I2C: start
-    xSemaphoreTake(i2c_mutex, portMAX_DELAY);
+  // CRITICAL SECTION I2C: start
+  xSemaphoreTake(i2c_mutex, portMAX_DELAY);
 
-    PCF8574::DigitalInput dra = pcf8574.digitalReadAll();
+  // setup pins
+  pcf8574.pinMode(P0, INPUT);
+  pcf8574.pinMode(P1, INPUT);
+  pcf8574.pinMode(P2, INPUT);
+  pcf8574.pinMode(P3, INPUT);
+  pcf8574.pinMode(P4, INPUT);
+  pcf8574.pinMode(P5, INPUT);
+  pcf8574.pinMode(P6, INPUT);
+  pcf8574.pinMode(P7, INPUT);
```

Comment:
Are they always inputs or is this just how you want them to be initialized?

Reply:
Initialize and always used as inputs only

Reply:
next feat: split for multiple instances of PCF8574

Reply:
Sounds good to me. Let's discuss this on Sunday :)